swiftsimio/accelerated.py

```diff
@@ -168,3 +168,317 @@ def index_dataset(handle: Dataset, mask_array: np.array) -> np.array:
+        a single dimension this is not used.
+
+    
+    Returns
+    -------
+
+    array: np.ndarray
+        Result from reading only the relevant values from ``handle``.
+    """
+
+    chunk_size = handle.chunks[0]
+
+    # Make array of chunk ranges
+    chunk_ranges = get_chunk_ranges(ranges, chunk_size, handle.shape[0])
+    chunk_range_size = np.diff(chunk_ranges).sum()
+
+    try:
+        output_shape = (chunk_range_size, output_shape[1])
+    except:
+        # Output shape is just a number, we have a 1D array.
+        output_shape = chunk_range_size
+
+    output = np.empty(output_shape, dtype=output_type)
+    already_read = 0
+    handle_multidim = handle.ndim > 1
+
+    for bounds in chunk_ranges:
+        read_start, read_end = bounds
+        if read_end == read_start:
+            continue
+
+        # Because we read inclusively
+        size_of_range = read_end - read_start
+
+        # Construct selectors so we can use read_direct to prevent creating
+        # copies of data from the hdf5 file.
+        hdf5_read_sel = (
+            np.s_[read_start:read_end, columns]
+            if handle_multidim
+            else np.s_[read_start:read_end]
+        )
+
+        output_dest_sel = np.s_[already_read : size_of_range + already_read]
+
+        handle.read_direct(output, source_sel=hdf5_read_sel, dest_sel=output_dest_sel)
+
+        already_read += size_of_range
+
+    if handle.chunks is not None:
+        return extract_ranges_from_chunks(output, chunk_ranges, ranges)
+    else:
+        return output
+
+
+def read_ranges_from_file(
+    handle: Dataset,
+    ranges: np.ndarray,
+    output_shape: Tuple,
+    output_type: type = np.float64,
+    columns: np.lib.index_tricks.IndexExpression = np.s_[:],
+) -> np.array:
+    """
+    Wrapper function to correctly select which version of read_ranges_from_file
+    should be used
+
+    Parameters
+    ----------
+
+    handle: Dataset
+        HDF5 dataset to slice data from
+
+    ranges: np.ndarray
+        Array of ranges (see :func:`ranges_from_array`)
+
+    output_shape: Tuple
+        Resultant shape of output. 
+    
+    output_type: type, optional
+        ``numpy`` type of output elements. If not supplied, we assume ``np.float64``.
+
+    columns: np.lib.index_tricks.IndexExpression, optional
+        Selector for columns if using a multi-dimensional array. If the array is only
+        a single dimension this is not used.
+
+    
+    Returns
+    -------
+
+    array: np.ndarray
+        Result from reading only the relevant values from ``handle``.
+
+    See Also
+    --------
+    read_ranges_from_file_chunked: reads data within specified ranges for chunked hdf5 file
+    read_ranges_from_file_unchunked: reads data within specified ranges for unchunked hdf5 file
+    """
+
+    average_range_size = np.diff(ranges).sum()
+    read_ranges = (
+        read_ranges_from_file_chunked
+        if handle.chunks is not None and average_range_size < 5e5
```

Comment:
Why is this the case?

Also you have not computed an average here. You've computed a total number of particles in the variable `average_range_size`.

Reply:
In the bandwidth plots I did previously the "old" read_ranges_from_file, which is now read_ranges_from_file_unchunked, got significantly slower at small chunk sizes, but was faster for chunks greater than 1e6. So we only really want to use read_ranges_from_file_chunked when the ranges are small. 

Having said that the difference is likely to be small as the performance of the chunked version is ok even at large range sizes, so I can remove it if you prefer.

Reply:
Don't know what I was thinking when calculating the average though... Fixed it now.

Reply:
Final thing to do before merging is to add a comment about this magic number

Reply:
Done.